slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/DbPruner.java

```diff
@@ -0,0 +1,148 @@
+
+import tech.pegasys.web3signer.slashingprotection.DbLocker.LockType;
+import tech.pegasys.web3signer.slashingprotection.dao.LowWatermarkDao;
+import tech.pegasys.web3signer.slashingprotection.dao.SignedAttestation;
+import tech.pegasys.web3signer.slashingprotection.dao.SignedAttestationsDao;
+import tech.pegasys.web3signer.slashingprotection.dao.SignedBlock;
+import tech.pegasys.web3signer.slashingprotection.dao.SignedBlocksDao;
+import tech.pegasys.web3signer.slashingprotection.dao.SigningWatermark;
+
+import java.util.Optional;
+
+import org.apache.tuweni.units.bigints.UInt64;
+import org.apache.tuweni.units.bigints.UInt64s;
+import org.jdbi.v3.core.Handle;
+import org.jdbi.v3.core.Jdbi;
+
+public class DbPruner {
+  private final Jdbi jdbi;
+  private final SignedBlocksDao signedBlocksDao;
+  private final SignedAttestationsDao signedAttestationsDao;
+  private final LowWatermarkDao lowWatermarkDao;
+
+  public DbPruner(
+      final Jdbi jdbi,
+      final SignedBlocksDao signedBlocksDao,
+      final SignedAttestationsDao signedAttestationsDao,
+      final LowWatermarkDao lowWatermarkDao) {
+    this.jdbi = jdbi;
+    this.signedBlocksDao = signedBlocksDao;
+    this.signedAttestationsDao = signedAttestationsDao;
+    this.lowWatermarkDao = lowWatermarkDao;
+  }
+
+  public void pruneForValidator(
+      final int validatorId, final long epochsToKeep, final long slotsPerEpoch) {
+    checkArgument(
+        epochsToKeep > 0, "epochsToKeep must be a positive value, but was %s", epochsToKeep);
+    checkArgument(
+        slotsPerEpoch > 0, "slotsPerEpoch must be a positive value, but was %", slotsPerEpoch);
+    final long slotsToKeep = epochsToKeep * slotsPerEpoch;
+    pruneBlocks(validatorId, slotsToKeep);
+    pruneAttestations(validatorId, epochsToKeep);
+  }
+
+  private void pruneBlocks(final int validatorId, final long slotsToKeep) {
+    final boolean hasWatermark =
+        jdbi.inTransaction(
+            READ_UNCOMMITTED,
+            h -> {
+              lockForValidator(h, LockType.BLOCK, validatorId);
+              return moveWatermarkForBlock(validatorId, slotsToKeep, h);
+            });
+
+    if (hasWatermark) {
+      jdbi.useTransaction(
+          READ_UNCOMMITTED, h -> signedBlocksDao.deleteBlocksBelowWatermark(h, validatorId));
+    }
+  }
+
+  private boolean moveWatermarkForBlock(
+      final int validatorId, final long slotsToKeep, final Handle handle) {
+    final Optional<UInt64> watermarkSlot =
+        lowWatermarkDao
+            .findLowWatermarkForValidator(handle, validatorId)
+            .map(SigningWatermark::getSlot);
+    final Optional<UInt64> mostRecentSlot = signedBlocksDao.findMaxSlot(handle, validatorId);
+    final Optional<UInt64> slotWatermark =
+        calculateWatermark(slotsToKeep, mostRecentSlot, watermarkSlot);
+    final Optional<SignedBlock> watermark =
+        slotWatermark.flatMap(
+            w -> signedBlocksDao.findNearestBlockWithSlot(handle, validatorId, w));
```

Comment:
nit: findFirstBlockNewerThanSlot(x)?

Reply:
tbh - if slotWatermark doesn't exist, then no need to continue? [no need to change what's here, but just a thought]